### ISSUE TITLE
Allow local vars

### ISSUE THREAD

**Issue by hypotext** [penrose/penrose#372]
Instead of requiring the user to declare variables like `LOCAL.x = ...` and refer to them as `LOCAL.x`...

allow the user to declare variables like `var x = ...` and refer to it as simply `x` (implicitly in the local scope).

**Comment by keenancrane** [comment]
To give a concrete example, for the `hyperbolic` domain we wanted to implement the computation

```
u = (h.center - H.center)/norm( h.center - H.center)
```

i.e., compute some vector by normalizing the difference of two vectors.  Currently the best we can do is this (!):

```
   LOCAL.ux = LOCAL.c1x - LOCAL.c0x
   LOCAL.uy = LOCAL.c1y - LOCAL.c0y
   LOCAL.m = sqrt( LOCAL.ux*LOCAL.ux + LOCAL.uy*LOCAL.uy )
   LOCAL.vx = LOCAL.ux / LOCAL.m
   LOCAL.vy = LOCAL.uy / LOCAL.m
```

Part of making this code simpler is better handling of vector types, but it would already be very helpful to get rid of the `LOCAL.`s everywhere, turning it into this:

```
   local ux = c1x - c0x
   local uy = c1y - c0y
   local m = sqrt( ux*ux + uy*uy )
   local vx = ux / m
   local vy = uy / m
```

**Comment by keenancrane** [comment]
Some design questions:

- What keyword(s) should we use? `local`? `val`? `var`? `let`? `const`? …. Or should there even be a keyword for local variables?  Having a keyword may make it easier to see where something was declared vs. assigned to.
- Should variables all be immutable?
- Should we have local scoping rules even within blocks?  E.g., variables that get declared inside a shape?  (Probably not, since we don't have nesting.)
- Should we have types?  E.g., `scalar` vs. `vector2` vs. `vector3` (just for readability)?  Note that we have a pretty domain-specific scope here, so it's not like we're gonna go wild with all sorts of crazy/custom types.

Probably not a bad idea to look at GLSL:

- Wikipedia page: https://en.wikipedia.org/wiki/OpenGL_Shading_Language
- Specification: https://www.khronos.org/registry/OpenGL/specs/gl/GLSLangSpec.4.60.pdf

**Comment by wodeni** [comment]
Addressed by #403.
